sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -2586,4 +2586,15 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     }
     assert(!jobStarted.get(), "Command should not trigger a Spark job.")
   }
+
+  test("SPARK-14471 Aliases in SELECT could be used in GROUP BY") {
+    Seq((1, "a", 0), (2, "a", 1), (1, "a", 2)).toDF("k1", "k2", "v")
+      .createOrReplaceTempView("t")
+    checkAnswer(
+      sql("SELECT k1 AS key1, k2 AS key2, SUM(v) FROM t GROUP BY key1, key2"),
+      Row(1, "a", 2) :: Row(2, "a", 1) :: Nil)
+    checkAnswer(
+      sql("SELECT k1 AS key1, key1 + 1 AS key2, COUNT(1) FROM t GROUP BY key1, key2"),
+      Row(1, 2, 2) :: Row(2, 3, 1) :: Nil)
+  }
```

Comment:
I think the target of this pr just replaces unresolved exprs in grouping keys with resolved ones by using resolved exprs in SELECT clauses. So, I think we'd better to discuss the name collision case you described in follow-up other tickets.